### PR TITLE
Fix #10108: Distinguish between soft and hard union types

### LINKED ISSUE #10108: Type inference for union types is surprising
## Minimized example

```Scala
object Test1:
  val x: Int | String = 1
  val y = x
  val z: Int | String = y
```

## Output

```
4 |  val z: Int | String = y
  |                        ^
  |                        Found:    (Test1.y : Any)
  |                        Required: Int | String
```

The problem is that a union type is widened at each inference step. Hence the declared type `Int | String` of `x` gets lost in the type of `y`. 

This widening can lead to mysterious behavior in the type checker. Here's an example:

```scala
object Test2:
  type Sig = Int | String
  def consistent(x: Sig, y: Sig): Boolean = x == y

  def consistentLists(xs: List[Sig], ys: List[Sig]): Boolean =
       xs.corresponds(ys)(consistent)        // OK
    || xs.corresponds(ys)(consistent(_, _))  // error, found: Any, required: Int | String
```

The first `corresponds` tests succeeds, but if we eta-expand the argument to `consistent(_, _)` it fails.
 
## Expectation

I hope we can come up with a more predictable type inference algorithm.

### PR DESCRIPTION
**Pull request by odersky**
The new system widens only soft union types using a join. If a hard union type appears in a soft one, it is dissolved together with the soft union type. Unions with Null continue to be treated specially.

### CODE REVIEW

**Comment by odersky** [comment]
@liufengyun We get quite a few different check files in pattern matching. One problem is that an extra empty line got added in several of them. There are also two completely different outputs for tests involving hard union types. In one case, the check file is now 8MB long! I updated the check files in 248f9eb. Can you take a look at the failures, and possibly also revert 248f9eb?

**Comment by liufengyun** [comment]
OK, I'll have a look and fix the exhaustivity check problems.

File: compiler/test/dotty/tools/dotc/transform/PatmatExhaustivityTest.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ import org.junit.Test
 class PatmatExhaustivityTest {
   val testsDir = "tests/patmat"
   // stop-after: patmatexhaust-huge.scala crash compiler
-  val options = List("-color:never", "-Ystop-after:crossCast", "-Ycheck-all-patmat", "-classpath", TestConfiguration.basicClasspath)
+  val options = List("-color:never", "-Ystop-after:explicitSelf", "-Ycheck-all-patmat", "-classpath", TestConfiguration.basicClasspath)
```
Comment (odersky) [inline compiler/test/dotty/tools/dotc/transform/PatmatExhaustivityTest.scala:21]: That explains the mystery of the added lines, I suppose? Good catch!

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 2]
```diff
@@ -150,19 +150,22 @@ trait SpaceLogic {
...
-        case ss  =>
-          ss.foldLeft(List[Prod]()) { (acc, flat) =>
-            if (acc.isEmpty) flat.map(s => Prod(tp, fun, s :: Nil, full))
-            else for (Prod(tp, fun, ss, full) <- acc; s <- flat) yield Prod(tp, fun, ss :+ s, full)
-          }
+      val ss = LazyList(spaces: _*).map(flatten)
+
+      ss.foldLeft(LazyList(Nil : List[Space])) { (acc, flat) =>
+        for { sps <- acc; s <- flat }
+        yield sps :+ s
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:159]: Nice!

**Comment by odersky** [comment]
Thanks for the fixes Fengyun! Your changes LGTM. Can you do the review for the rest?

File: compiler/test/dotty/tools/dotc/transform/PatmatExhaustivityTest.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/test/dotty/tools/dotc/transform/PatmatExhaustivityTest.scala:21]: Yes, most failures come from this change. The test case `i8922c.scala` and `i4030.scala` are due to type inference change.

File: tests/patmat/i4030.scala [hunk 3]
```diff
@@ -9,5 +9,6 @@ object TestGADT {
   def f[A <: Seq[_], B, Foo >: A => B](v: Root[Foo], u: Root[Foo]) = (v, u) match {
     case (C3(), C3()) =>
   }
-  f(C3[Seq[_], Long](), C4[Seq[_], Long]())
+  // The following line no longer type checks
+  // f(C3[Seq[_], Long](), C4[Seq[_], Long]())
```
Comment (liufengyun) [inline tests/patmat/i4030.scala:13]: This test case shows that this PR improves over the previous handling of union types 👍

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**